### ISSUE TITLE
Some questions about the Match R-CNN in the paper

### ISSUE THREAD

**Issue by xwjabc** [switchablenorms/DeepFashion2#17]
Hi! I have some questions about the design of Match R-CNN:
1) What are the coefficient lambda_1 ~ lambda_5 in the loss function L?
2) I would like to know the details to create pairs in training stage.
My understanding: Assume we have batch_size=16. Thus, we will have 8 instance pairs where 4 are positive pairs and 4 are negative pairs:
- To form the positive pairs, we sample one user instance (source = 'user') with tuple (pair_id, style_id) from all images (style_id != 0) and then sample one commercial instance (source = 'shop') with the **same** tuple (pair_id, style_id). 
- To form the negative pairs, we sample one another user instance (source = 'user') with tuple (pair_id, style_id) from all images (style_id != 0) and then sample one commercial instance (source = 'shop') with the **different** tuple (pair_id', style_id').
Afterwards, we train the Match R-CNN with bbox / mask on **all proposals** and use the **best  one** proposal (i.e. the proposal has the highest IoU with the ground-truth bbox of the instance in the pair) assigned to each instance in the pairs to obtain the feature and feed it into matching network.
Q1: Is my understanding correct? Do we use the same number of positive pairs and negative pairs? 
Q2: In bbox / mask / keypoint training of Mask R-CNN/Match R-CNN, we will optimize the loss on all selected proposals. However, for the matching network, do we use **one** proposal of each instance in the pairs or **all** proposals of each instance in the pairs? If we use all proposals, there will be many combinations of proposals for matching network (count: #proposals of user instance * #proposals of commercial instance).

**Comment by geyuying** [comment]
1. Match R-CNN is an end-to-end framework and learning all four tasks jointly enables a unified system
to efficiently predict all outputs simultaneously. But through experiments, we find that for example, loss of keypoint branch reduces the performance of segmentation slightly. So for different task, we assign different coefficient lambda_1 ~ lambda_5 to achieve better results.
2. As for creating pairs in training stage, for batch_size=16, we have 8 images from consumers and select 8 corresponding images from shops. (Image from consumers and its corresponding image from shops means that the two images have at least one positive pair.) 
When training, for each image, we select fixed number of proposals(i.e 8). We assume that  in each image there are no more than 2 instances whose 'style' is >0 and for each of these instance, we select 8 proposals(if there is only 1 instance) or 4 proposals( if there are 2 instances) according to the IoU with the ground-truth bbox (i.e the IoU is above some certain thresh). Then we have 8x8=64 instances from shops and 8x8 = 64 instances from users and instances with the same tuple  (pair_id, style_id) construct positive pairs, otherwise they are negative pairs.

**Comment by xwjabc** [comment]
Thank you for your reply! I still have a bit confusion:
1. For the coefficient lambda_1 ~ lambda_5, can you provide a set of values of those coefficients (e.g. pose+class because Table 5 shows that it has the best retrieval result)? I am currently trying to implement the Match R-CNN and it would be really helpful if there is some coefficient setting for reference.
2. So it seems we always have 64 instances (proposals) from consumers and 64 instances (proposals) from shops. Thus, if we consider all possible combinations, there will be 4096 pairs. Among the 4096 pairs, there will be ~512 (=8*8*8 if we assume each image has one 'style>0' instance) positive pairs and ~3584 (=4096-512) negative pairs. Is my understanding correct? Besides, there will be more negative pairs than positive ones. Should we balance the effect from positive and negative pairs by using different weight on the positive pairs' loss and negative pairs' loss?

**Comment by geyuying** [comment]
1. You can refer to issue 14 for details about retrieval part in Match R-CNN.
2.Your understanding is correct. You can also reduce the number of pairs.(i.e construct 1 pair with one instance in shop image and one of the 8 instances in user image instead of 8 pairs). We didn't use  different weights on the positive pairs' loss and negative pairs' loss, but it may perform better and you can have a try.

**Comment by xwjabc** [comment]
Thank you for your reply! I read issue 14, but there is no value for the coefficient lambda_1 ~ lambda_5. Could you provide the values of the coefficients? Thanks!

**Comment by xwjabc** [comment]
1. Recently I am implementing the Match R-CNN framework. I found in my PyTorch code of Mask R-CNN, the DistributedDataParallel is used for GPU parallelization. Thus, if we have batch_size=16 for 8 GPUs, each GPU will only have 2 images. In my understanding, the 2 images include 1 user image and 1 shop image. According to your reply above, there will be 8 user proposals and 8 shop proposals, which creates 8*8=64 positive pairs. However, it seems there are no negative pairs. Then how to get the negative pairs?

Another possibility is, after obtaining the proposals, we use all_gather to transfer the proposals to every GPU (which means each GPU will receive all proposals from other GPUs). Thus, we can use the way we previously discussed (64 user proposals, 64 shop proposals -> 512 positive pairs and 3584 negative pairs). It will consume more GPU memory and brings about some communication overhead.

2. By the way, I have another confusion: In Mask R-CNN, there will be 512 proposals for the RoI heads (ideally 128 of 512 are foreground and 384 of 512 are background, but in actual training 30~40 are foreground and the rest are background). Thus, as you mentioned, 8 proposals (determined by the IoU above some certain thresh, in paper it is 0.7) are used. I wonder if there is a sampling mechanism from 30~40 to the 8 proposals (e.g. random sampling / choose top-8 IoU proposals / etc).

3. One last question, the paper mentioned there are 12 epochs for training. I wonder what is the definition of "epochs" here: Does it mean the number of user images (with 1~2 style>0 instances and there is at least 1 corresponding shop image having at least 1 instance of same pair_id and style)?

**Comment by xwjabc** [comment]
Besides, in the evaluation, I wonder if Match R-CNN use the in-class setting or not: For in-class evaluation, we only compare the query instances and the gallery instances with the same class and choose top-k. For non-in-class evaluation, we compare the query instances with all gallery instances. Since the in-class way will usually perform better than the non-in-class way, I wonder which one is used in the retrieval evaluation of Match R-CNN.

**Comment by AlvinAi96** [comment]
> Besides, in the evaluation, I wonder if Match R-CNN use the in-class setting or not: For in-class evaluation, we only compare the query instances and the gallery instances with the same class and choose top-k. For non-in-class evaluation, we compare the query instances with all gallery instances. Since the in-class way will usually perform better than the non-in-class way, I wonder which one is used in the retrieval evaluation of Match R-CNN.

Hi, Thanks for many insightful views, I am working on remake the code on the paper but I am trapped into the pair loss updating problem. I input 8 images on a GPU as acnhor images, then I get 8 positive images and 8 negative images from the dataset. SInce I am only use one top-score proposal from one image, the positive pairs has 8 number sand the negative pairs has 8 numbers too. I use sigmoid as the match network output and BCEloss to compute the loss among those pairs. But the loss of positive pairs and the loss of negative pairs are almost the same.  Each batch i compute 16 pair -losses (8 postive-pair losses and 8 negative-pair losses) and average them, then updating the gradient of the model. have you face a similar problem of the loss of postive and negative  pair are almost the same? how do you compute the loss and update them. Would you give a suggestion? Thanks.

**Comment by xwjabc** [comment]
> > Besides, in the evaluation, I wonder if Match R-CNN use the in-class setting or not: For in-class evaluation, we only compare the query instances and the gallery instances with the same class and choose top-k. For non-in-class evaluation, we compare the query instances with all gallery instances. Since the in-class way will usually perform better than the non-in-class way, I wonder which one is used in the retrieval evaluation of Match R-CNN.
> 
> Hi, Thanks for many insightful views, I am working on remake the code on the paper but I am trapped into the pair loss updating problem. I input 8 images on a GPU as acnhor images, then I get 8 positive images and 8 negative images from the dataset. SInce I am only use one top-score proposal from one image, the positive pairs has 8 number sand the negative pairs has 8 numbers too. I use sigmoid as the match network output and BCEloss to compute the loss among those pairs. But the loss of positive pairs and the loss of negative pairs are almost the same. Each batch i compute 16 pair -losses (8 postive-pair losses and 8 negative-pair losses) and average them, then updating the gradient of the model. have you face a similar problem of the loss of postive and negative pair are almost the same? how do you compute the loss and update them. Would you give a suggestion? Thanks.

Hi, I am sorry that it has been a while since I re-implemented the Match R-CNN. Thus, I need to re-read my code and answer your questions later. I am also preparing to release my re-implementation later and you can check the details.

**Comment by AlvinAi96** [comment]
> > > Besides, in the evaluation, I wonder if Match R-CNN use the in-class setting or not: For in-class evaluation, we only compare the query instances and the gallery instances with the same class and choose top-k. For non-in-class evaluation, we compare the query instances with all gallery instances. Since the in-class way will usually perform better than the non-in-class way, I wonder which one is used in the retrieval evaluation of Match R-CNN.
> > 
> > 
> > Hi, Thanks for many insightful views, I am working on remake the code on the paper but I am trapped into the pair loss updating problem. I input 8 images on a GPU as acnhor images, then I get 8 positive images and 8 negative images from the dataset. SInce I am only use one top-score proposal from one image, the positive pairs has 8 number sand the negative pairs has 8 numbers too. I use sigmoid as the match network output and BCEloss to compute the loss among those pairs. But the loss of positive pairs and the loss of negative pairs are almost the same. Each batch i compute 16 pair -losses (8 postive-pair losses and 8 negative-pair losses) and average them, then updating the gradient of the model. have you face a similar problem of the loss of postive and negative pair are almost the same? how do you compute the loss and update them. Would you give a suggestion? Thanks.
> 
> Hi, I am sorry that it has been a while since I re-implemented the Match R-CNN. Thus, I need to re-read my code and answer your questions later. I am also preparing to release my re-implementation later and you can check the details.

Thanks. would you like to inform my through email If you release your code? My email: alvinai9603@outlook.com. I am looking forward to it.

**Comment by VyBui** [comment]
> > > Besides, in the evaluation, I wonder if Match R-CNN use the in-class setting or not: For in-class evaluation, we only compare the query instances and the gallery instances with the same class and choose top-k. For non-in-class evaluation, we compare the query instances with all gallery instances. Since the in-class way will usually perform better than the non-in-class way, I wonder which one is used in the retrieval evaluation of Match R-CNN.
> > 
> > 
> > Hi, Thanks for many insightful views, I am working on remake the code on the paper but I am trapped into the pair loss updating problem. I input 8 images on a GPU as acnhor images, then I get 8 positive images and 8 negative images from the dataset. SInce I am only use one top-score proposal from one image, the positive pairs has 8 number sand the negative pairs has 8 numbers too. I use sigmoid as the match network output and BCEloss to compute the loss among those pairs. But the loss of positive pairs and the loss of negative pairs are almost the same. Each batch i compute 16 pair -losses (8 postive-pair losses and 8 negative-pair losses) and average them, then updating the gradient of the model. have you face a similar problem of the loss of postive and negative pair are almost the same? how do you compute the loss and update them. Would you give a suggestion? Thanks.
> 
> Hi, I am sorry that it has been a while since I re-implemented the Match R-CNN. Thus, I need to re-read my code and answer your questions later. I am also preparing to release my re-implementation later and you can check the details.

Hello @xwjabc, will the code be available soon?
